Авторизация-бекенд: UUID для сущностей PG + настроил авто генерацию на стороне PG

ZakharMolchanov opened the pull request.

ZakharMolchanov committed:
Прописал всем id тип UUID + в сущностях, также в PG должно подхватиться(менял ddl)

ZakharMolchanov committed:
Добавлены роуты для регистрации и авторизации.  Для аутентификации ..

ZakharMolchanov commented:
Чет чел ругается, из-за файлов. Чекните пж. При авторизации и регистрации выдается в куки JWT токен, шифруется приватником, в пейлоаде передает пока login( токен живет час) .Основная задача jwt токена, это аутентификация( публчиный ключ сгенерен из приватника, можно чекать подпись jwt публичным).

disk0Dancer commented on package.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "dependencies": {
+    "bcrypt": "^5.1.1",
+    "express": "^4.19.2",
+    "express-validator": "^7.1.0",
+    "fs": "^0.0.1-security",
+    "jsonwebtoken": "^9.0.2",
+    "uuid": "^10.0.0"
+  }
+}
```
Норм, что packge.json появился там, где у нас нет проекта? 
я, кажется уже у себя удалял пару раз его, хз нужен он нет?

ZakharMolchanov replied:
Эх, устанавливал пакеты в открытой директории train-platform, а не в backend. Этот package.json не нужен.

disk0Dancer replied:
Панимаю, убери пж потом в след комите, перед слияним.

disk0Dancer commented on backend/src/controllers/AuthController.ts:
```diff
@@ -0,0 +1,150 @@
+      }
+
+      // Хэширование пароля
+      const hashedPassword = await bcrypt.hash(password, 10);
```
Нет проблем с пакетом при сборке через докер?

ZakharMolchanov replied:
Проблем нет, контейнер бэкенда стартует нормально. В БД записывает корректно 
![image](https://github.com/TeamHSE/train-platform/assets/92994381/e9855256-ad72-4c95-96e2-fdbcd868abcd)

disk0Dancer replied:
У меня такая фигня выпала 
<img width="564" alt="image" src="https://github.com/TeamHSE/train-platform/assets/89835485/42e1d0fc-4970-47c0-9f5a-df6d85814bac">

ZakharMolchanov replied:
Эта проблема из-за коммита https://github.com/TeamHSE/train-platform/commit/3a0594c3f95e5d4934e83afe647f954b0f9e6bab

disk0Dancer commented on backend/src/controllers/AuthController.ts:
```diff
@@ -0,0 +1,150 @@
+    check("login").isEmail().withMessage("Неверный формат почты!"),
+    check("password")
+      .isLength({ min: 6 })
+      .withMessage("Пароль слишком короткий!"),
+    check("username")
+      .not()
+      .isEmpty()
+      .withMessage("Имя пользователя не может быть пустым!"),
+    check("firstName").not().isEmpty().withMessage("Имя не может быть пустым!"),
+    check("lastName")
+      .not()
+      .isEmpty()
+      .withMessage("Фамилия не может быть пустой!"),
+
+    async (req: Request, res: Response) => {
+      const errors = validationResult(req);
+      if (!errors.isEmpty()) {
+        return res.status(422).json({ errors: errors.array() });
+      }
+
+      const { login, password, username, firstName, lastName } = req.body;
+
+      // Проверка занятости логина
+      const userRepository = AppDataSource.getRepository(User);
+      const existingEmail = await userRepository.findOne({
+        where: { email: login },
+      });
+      if (existingEmail) {
+        return res
+          .status(409)
+          .json({ message: "Данный Email уже используется" });
+      }
+
+      // Проверка занятости имени пользователя
+      const existingUsername = await userRepository.findOne({
+        where: { username },
+      });
+      if (existingUsername) {
+        return res.status(409).json({ message: "Имя пользователя уже занято" });
+      }
```
Мб стоит извлечь код валидации запроса в отдельную функцию?
Будет лучше смотреться, как мне кажется, да и изменения легче будет вносить.

ZakharMolchanov replied:
Да, согласен. Исправлю

disk0Dancer commented on backend/src/controllers/AuthController.ts:
```diff
@@ -0,0 +1,150 @@
+
+      try {
+        // Чтение приватного ключа из файла
+        const privateKey = fs.readFileSync('private.pem', 'utf8');
```
мб стоит загрузить ключ в Config? и читать его из enviroment?

ZakharMolchanov replied:
Я чекал в инете, пишут, что публичный ключ можно в конфиг вынести, а приватный лучше хранить в файле. Но в целом, если проект учебный, то можно вынести все в конфиг

disk0Dancer replied:
Сделай метод для обьекта конфиг, чтобы он просто читал из файла.

Типо доступ в файловой системы из другого метода просто, не из обработчика.

ZakharMolchanov replied:
А окей, сделаю

disk0Dancer commented on backend/src/routes.ts:
```diff
@@ -2,11 +2,13 @@
   app.get("/", RootController);
-
+  app.post("/api/auth/register", AuthController.register);
+  app.post("/api/auth/login", AuthController.login);
```
Как я понял, предупреждение о том, что при обработке запроса мы лезем в файловую систему, каждый запрос обрабатывается асинхронно и хз как поведет себя программа в стрессе, наверное лучше убрать это дело в конфиг, и загружать через конфиг прикатный ключ.

Можно сделать метод, чтобы не хранить все время ключ в переменной, либо поле и пофиг.

ZakharMolchanov commented:
Прошу не чекать( надо бы его отменить) коммит #3a0594c3f95e5d4934e83afe647f954b0f9e6bab, неверно указаны типы данных

ZakharMolchanov committed:
feat(database): Change id column type from integer to uuid

ZakharMolchanov committed:
feat(database): Change id column type from integer to uuid

ZakharMolchanov committed:
Merge remote-tracking branch 'origin/feat/auth' into feat/auth

# Conflicts:
#	backend/src/entity/Course.ts
#	backend/src/entity/CoursesModules.ts
#	backend/src/entity/CoursesStep.ts
#	backend/src/entity/CoursesViewers.ts
#	backend/src/entity/UsersActivity.ts
#	backend/src/entity/UsersRoles.ts

ZakharMolchanov committed:
feat(backend): Moved public and private key reading function into config
Separated registration and authentication routes into two separate routes. Also added validation function.Addressed review comments.

ZakharMolchanov committed:
feat(backend): Removed packages located above backend

AleksanderNekr commented on backend/src/controllers/AuthControllerLogin.ts:
```diff
@@ -0,0 +1,52 @@
+import fs from 'fs';
```
Лишняя зависимость здесь, проверь

AleksanderNekr commented on backend/src/controllers/AuthControllerLogin.ts:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    try {
+      const privateKey = config.readPrivateKey('path/to/private.pem'); // Update with the actual path
```
Тут что-то не так с путем)

AleksanderNekr commented on backend/src/controllers/AuthControllerLogin.ts:
```diff
@@ -0,0 +1,52 @@
+      return res.status(400).json({ message: "Неверный логин или пароль" });
+    }
+
+    const isPasswordValid = await bcrypt.compare(password, user.password);
```
Здесь авейт лишний вроде как

AleksanderNekr commented on backend/src/controllers/AuthControllerRegister.ts:
```diff
@@ -0,0 +1,72 @@
+import jwt from "jsonwebtoken";
+import { userValidationRules } from "../middlewares/validation";
+import { config } from "../config/Config";
+import * as fs from 'fs';
```
Здесь фс тоже лишний по идее

AleksanderNekr commented on backend/src/entity/Course.ts:
```diff
@@ -1,9 +1,10 @@
 import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";
 
+
 @Entity("courses")
```
Слишком много переносов строки – отформатируй код пж

AleksanderNekr commented on backend/src/entity/CoursesModules.ts:
```diff
@@ -1,12 +1,14 @@
 
+
+
 @Entity("courses_modules")
```
Тоже переносов много

AleksanderNekr commented on backend/src/entity/CoursesStep.ts:
```diff
@@ -1,12 +1,14 @@
 import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";
 
+
+
```
Переносы

AleksanderNekr commented on backend/src/entity/CoursesViewers.ts:
```diff
@@ -1,16 +1,18 @@
 import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";
 
+
+
```
Переносы)

AleksanderNekr commented on backend/src/entity/User.ts:
```diff
@@ -1,10 +1,11 @@
-import { UUID } from "crypto";
+
+
 
```
Переносов много

ZakharMolchanov committed:
feat(backend): delete package-lock in backend

AleksanderNekr commented on backend/src/entity/UsersRoles.ts:
```diff
@@ -1,9 +1,10 @@
 import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";
 
+
```
Тоже перенос лишний

ZakharMolchanov committed:
feat(backend): Corrected comments from PR

AleksanderNekr reviewed (changes requested):
Настроить CORS-политику для URL фронта: (localhost:3000, лучше брать из конфига url) на доступ к конечным точкам и пересылки credentials

AleksanderNekr commented:
[] Добавить генерацию RSA-ключей в Dockerfile
[] Настроить CORS

ZakharMolchanov committed:
feat(backend): update dockerfile, generate public key and private key

ZakharMolchanov committed:
feat(backend): add cors policy

AleksanderNekr commented:
От меня ОК, @disk0Dancer, посмотри пж

disk0Dancer commented on backend/src/config/Config.ts:
```diff
@@ -1,10 +1,48 @@
+const publicKeyPath = path.resolve('src/public.pem');
+
+
+const frontendURL = "http://localhost:3001";
```
- лучше убрать в enviroment
- добавить в compose.yml

disk0Dancer commented on backend/src/config/Config.ts:
```diff
@@ -1,10 +1,48 @@
-export const config = new Config();
-config.dbHost = process.env.DB_HOST;
+// Путь к файлам ключей
+const privateKeyPath = path.resolve('src/private.pem');
+const publicKeyPath = path.resolve('src/public.pem');
```
перемести файлы в ```backend/src/config```, думаю можно инициализировать прямо к конструкторе.
Только проверь, что .dockerignore не скрывает эти файлы

AleksanderNekr replied:
@disk0Dancer, кстати тут решили, что файлы ключей совать в гит не будем, так как это чувствительная информация, в докерфайле бэка добавлена генерация ключей

Чтобы тебе их получить для локальной разработки без докера, можешь у Захара их попросить, или сгенерить

disk0Dancer replied:
Хорошее предложение, добавьте инструкцию к запуску(генерации) пж

ZakharMolchanov replied:
в енвайромент закинул пути
генерятся ключи в src/config

ZakharMolchanov committed:
feat(backen-auth): fix problem with login

ZakharMolchanov committed:
feat(backend): display the key paths in dockercompose
